### PR TITLE
Migrate temp_repo.go to use git.NewCommand 

### PR DESCRIPTION
**Pull request by zeripath**
This PR migrates temp_repo.go to use git.NewCommand instead creating processes by itself - this fixes the problem underlying PR #8905.

There are other places that run git outside of the controlled locale defined in #8548 but temp_repo.go is the only cause of failure of local testing in cases where English is not the default - implying that error messages from those other commands are not interpreted.

Replaces #8905

### CODE REVIEW

**Comment by zeripath** [comment]
I would in particular appreciate reviews of the changes to modules/git/command.go - this may not be the most sensible way of causing a callback between start and wait.

**Review by 6543 — APPROVED** [review]
if fix problems at `make test` with locales
and we are going to have only one package who makes git processes ... :+1:

**Review by lunny — APPROVED** [review]

File: modules/repofiles/temp_repo.go [hunk 1]
```diff
@@ -287,65 +222,62 @@ func (t *TemporaryUploadRepository) CommitTree(author, committer *models.User, t
...
+		if finalErr != nil {
+			log.Error("Unable to ParsePatch in temporary repo %s (%s). Error: %v", t.repo.FullName(), t.basePath, finalErr)
+			return nil, finalErr
+		}
+		log.Error("Unable to run diff-index pipeline in temporary repo %s (%s). Error: %v\nStderr: %s",
+			t.repo.FullName(), t.basePath, err, stderr)
+		return nil, fmt.Errorf("Unable to run diff-index pipeline in temporary repo %s. Error: %v\nStderr: %s",
+			t.repo.FullName(), err, stderr)
 	}
+	_ = stdoutWriter.Close()
```
Comment (guillep2k) [inline modules/repofiles/temp_repo.go]: Isn't `stdoutReader.Close()` required here too?

Comment (guillep2k) [reply modules/repofiles/temp_repo.go]: In fact, there are several output paths. I'm struggling to visualize whether any of them leaks a handle. Is it possible to just use `defer stdoutWriter.Close()` and `defer stdoutReader.Close()`?

Comment (zeripath) [reply modules/repofiles/temp_repo.go]: This close is actually unnecessary as the writer is closed above. (it was left over from struggling to get it to work. 

To be bolts and braces though I'll add a defer close just after opening the things.

**Comment by zeripath** [comment]
This PR migrates temp_repo.go to use git.NewCommand instead creating processes by itself - this fixes the problem underlying PR #8905.

There are other places that run git outside of the controlled locale defined in #8548 but temp_repo.go is the only cause of failure of local testing in cases where English is not the default - implying that error messages from those other commands are not interpreted.

Replaces #8905

File: modules/repofiles/temp_repo.go [hunk 2]
```diff
@@ -287,64 +222,64 @@ func (t *TemporaryUploadRepository) CommitTree(author, committer *models.User, t
...
+
+	if err := git.NewCommand("diff-index", "--cached", "-p", "HEAD").
+		RunInDirTimeoutEnvFullPipelineFunc(nil, 30*time.Second, t.basePath, stdoutWriter, stderr, nil, func(ctx context.Context, cancel context.CancelFunc) {
+			_ = stdoutWriter.Close()
+			diff, finalErr = gitdiff.ParsePatch(setting.Git.MaxGitDiffLines, setting.Git.MaxGitDiffLineCharacters, setting.Git.MaxGitDiffFiles, stdoutReader)
+			if finalErr != nil {
+				log.Error("ParsePatch: %v", finalErr)
+				cancel()
+			}
+			_ = stdoutReader.Close()
```
Comment (guillep2k) [inline modules/repofiles/temp_repo.go:273]: Hummm... Now this and L267 are no longer required.  :sweat_smile:

Comment (zeripath) [reply modules/repofiles/temp_repo.go:273]: L267 is definitely still required. The process will hang without it.

L273 maybe isn't.

**Pull request merged**